src/_pytest/pathlib.py

```diff
@@ -397,3 +402,133 @@ def fnmatch_ex(pattern: str, path) -> bool:
+
+
+class ImportMode(Enum):
+    """Possible values for `mode` parameter of `import_module`"""
+
+    prepend = "prepend"
+    append = "append"
+    importlib = "importlib"
+
+
+class ImportPathMismatchError(ImportError):
+    """Raised on import_module() if there is a mismatch of __file__'s.
+
+    This can happen when `import_module` is given different filenames with the same basename,
+    when those filenames don't reside in packages.
+    """
+
+
+def import_module(
+    p: Union[str, py.path.local, Path],
+    *,
+    mode: Union[str, ImportMode] = ImportMode.prepend
+) -> ModuleType:
+    """
+    Imports and returns a module from the given path.
+
+    The import mechanism used is controlled by the `mode` parameter:
+
+    * `mode == ImportMode.prepend`: the directory containing the module (or package, taking
+      `__init__.py` files into account) will be put at the *start* of `sys.path`, if not
+      already in  `sys.path`, before being imported with `__import__.
+
+    * `mode == ImportMode.append`: same as `prepend`, but the directory will be appended
+      to the end of `sys.path`.
+
+    * `mode == ImportMode.importlib`: uses more fine control mechanisms provided by `importlib`
+      to import the module, which avoids having to use `__import__` and muck with `sys.path`
+      at all. It effectively allows having same-named test modules in different places.
+
+    :raise ImportPathMismatchError: if after importing the given `path` and the module `__file__`
+        are different. Only raised in `prepend` and `append` modes.
+    """
+    mode = ImportMode(mode)
+
+    path = Path(p)
+
+    if not path.exists():
+        raise ImportError(path)
+
+    if mode == ImportMode.importlib:
+        import importlib.util
+
+        modname = path.stem
+
+        for meta_importer in sys.meta_path:
+            spec = meta_importer.find_spec(modname, [str(path.parent)])
+            if spec is not None:
+                break
+        else:
+            spec = importlib.util.spec_from_file_location(modname, str(path))
+
+        if spec is None:
+            raise ImportError(
+                "Can't find module {} at location {}".format(modname, str(path))
+            )
+        mod = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(mod)  # type: ignore[union-attr]
+        return mod
+
+    pkgpath = resolve_package_path(path)
+    if pkgpath is not None:
+        pkgroot = pkgpath.parent
+        names = str(path.with_suffix("").relative_to(pkgroot)).split(os.path.sep)
+        if names[-1] == "__init__":
+            names.pop()
+        modname = ".".join(names)
+    else:
+        pkgroot = path.parent
+        modname = path.stem
+
+    if mode == ImportMode.append:
+        if str(pkgroot) not in sys.path:
+            sys.path.append(str(pkgroot))
+    else:
+        assert mode == ImportMode.prepend, "unexpected mode: {}".format(mode)
```

Comment:
Seems nice; I did implement it locally but I get:

```
src\_pytest\pathlib.py:495: error: Argument 1 to "assert_never" has incompatible type "ImportMode"; expected "NoReturn"  [arg-type]
Found 1 error in 1 file (checked 2 source files)
```

I implemented the solution from your original post verbatim, didn't really read the whole thread though.

Reply:
Here's a diff that works for me. (Again, it's ok if you prefer not to apply this)

```diff
diff --git a/src/_pytest/compat.py b/src/_pytest/compat.py
index 84f9609a7..c19beb505 100644
--- a/src/_pytest/compat.py
+++ b/src/_pytest/compat.py
@@ -33,6 +33,7 @@ else:
 
 
 if TYPE_CHECKING:
+    from typing import NoReturn
     from typing import Type
     from typing_extensions import Final
 
@@ -401,3 +402,38 @@ else:
     from collections import OrderedDict
 
     order_preserving_dict = OrderedDict
+
+
+# Perform exhaustiveness checking.
+#
+# Consider this example:
+#
+#     MyUnion = Union[int, str]
+#
+#     def handle(x: MyUnion) -> int {
+#         if isinstance(x, int):
+#             return 1
+#         elif isinstance(x, str):
+#             return 2
+#         else:
+#             raise Exception('unreachable')
+#
+# Now suppose we add a new variant:
+#
+#     MyUnion = Union[int, str, bytes]
+#
+# After doing this, we must remember ourselves to go and update the handle
+# function to handle the new variant.
+#
+# With `assert_never` we can do better:
+#
+#     // throw new Error('unreachable');
+#     return assert_never(x)
+#
+# Now, if we forget to handle the new variant, the type-checker will emit a
+# compile-time error, instead of the runtime error we would have gotten
+# previously.
+#
+# This also work for Enums (if you use `is` to compare) and Literals.
+def assert_never(value: NoReturn) -> NoReturn:
+    assert False, "Unhandled value: {} ({})".format(value, type(value).__name__)
diff --git a/src/_pytest/pathlib.py b/src/_pytest/pathlib.py
index 907e9223d..da7f0221c 100644
--- a/src/_pytest/pathlib.py
+++ b/src/_pytest/pathlib.py
@@ -23,6 +23,7 @@ from typing import Union
 
 import py
 
+from _pytest.compat import assert_never
 from _pytest.warning_types import PytestWarning
 
 if sys.version_info[:2] >= (3, 6):
@@ -453,7 +454,7 @@ def import_path(
     if not path.exists():
         raise ImportError(path)
 
-    if mode == ImportMode.importlib:
+    if mode is ImportMode.importlib:
         import importlib.util
 
         module_name = path.stem
@@ -484,13 +485,14 @@ def import_path(
         pkg_root = path.parent
         module_name = path.stem
 
-    if mode == ImportMode.append:
+    if mode is ImportMode.append:
         if str(pkg_root) not in sys.path:
             sys.path.append(str(pkg_root))
-    else:
-        assert mode == ImportMode.prepend, "unexpected mode: {}".format(mode)
+    elif mode is ImportMode.prepend:
         if str(pkg_root) != sys.path[0]:
             sys.path.insert(0, str(pkg_root))
+    else:
+        assert_never(mode)
 
     __import__(module_name)
 
```